platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/model/internal/deterministic/DeterministicHeartbeatScheduler.java

```diff
@@ -0,0 +1,110 @@
+import com.swirlds.common.wiring.model.internal.standard.HeartbeatTask;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+
+/**
+ * A deterministic implementation of a heartbeat scheduler.
+ */
+public class DeterministicHeartbeatScheduler extends AbstractHeartbeatScheduler {
+
+    /**
+     * This maps from the period of the heartbeat to the list of heartbeats that want to be notified with that period.
+     */
+    private final Map<Duration, List<HeartbeatTask>> heartbeatsByPeriod = new HashMap<>();
+
+    /**
+     * The timestamps of the last heartbeats sent for each group of heartbeats that subscribe to the same period.
+     */
+    private final Map<Duration, Instant> lastHeartbeats = new HashMap<>();
+
+    /**
+     * Constructor.
+     *
+     * @param model the wiring model containing this heartbeat scheduler
+     * @param time  provides wall clock time
+     * @param name  the name of the heartbeat scheduler
+     */
+    public DeterministicHeartbeatScheduler(
+            @NonNull final TraceableWiringModel model, @NonNull final Time time, @NonNull final String name) {
+        super(model, time, name);
+    }
+
+    /**
+     * Send out heartbeats based on the amount of time that has passed.
+     */
+    public void tick() {
+        if (!started) {
+            throw new IllegalStateException("Cannot tick the heartbeat before it has started");
+        }
+
+        final Instant currentTime = time.now();
+
+        for (final Entry<Duration, List<HeartbeatTask>> entry : heartbeatsByPeriod.entrySet()) {
+            final Duration period = entry.getKey();
+            final List<HeartbeatTask> tasksForPeriod = entry.getValue();
+            final Instant lastHeartbeat = lastHeartbeats.get(period);
+            final Duration timeSinceLastHeartbeat = Duration.between(lastHeartbeat, currentTime);
+            if (isGreaterThanOrEqualTo(timeSinceLastHeartbeat, period)) {
+                for (final HeartbeatTask task : tasksForPeriod) {
+                    task.run();
+                }
+                lastHeartbeats.put(period, currentTime);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void start() {
+        if (started) {
+            throw new IllegalStateException("Cannot start the heartbeat more than once");
+        }
+        started = true;
+
+        final Instant currentTime = time.now();
+        for (final HeartbeatTask task : tasks) {
+            final List<HeartbeatTask> tasksForPeriod =
+                    heartbeatsByPeriod.computeIfAbsent(task.getPeriod(), k -> new ArrayList<>());
+            tasksForPeriod.add(task);
+            lastHeartbeats.put(task.getPeriod(), currentTime);
+        }
```

Comment:
Would a failure in the subsequent processing after setting the started flag potentially cause the flag value to be wrong? E.g. if we set started to true and something happened in the computeIfAbsent that it threw an exception, would started still be correctly true? If that's the case, it's worth setting the started flag after the work has been done.

Reply:
If something throws inside the `computeIfAbsent` method, I'd expect for the exception to travel up the stack and to terminate whatever test is executing. In that case, I think it's not particularly important to worry about the state of the object that gets left behind, since presumably the object will have already been corrupted by the failure to build all heartbeat tasks.

My personal preference is to leave the `started = true` next to the other logic that is checking the started flag, but it's not that big of a change, so I moved it as requested.